src/compose/ComposeBox.js

```diff
@@ -217,6 +220,43 @@ class ComposeBox extends PureComponent<Props, State> {
     }
   };
 
+  insertAttachment = async (attachment: DocumentPickerResponse) => {
+    this.setState(({ numUploading }) => ({
+      numUploading: numUploading + 1,
+    }));
+
+    const { _, auth } = this.props;
+
+    const fileName = attachment.name ?? _('Attachment');
+    const placeholder = `[${_('Uploading {fileName}...', { fileName })}]()`;
+    this.insertMessageTextAtCursorPosition(`${placeholder}\n\n`);
+
+    let response = null;
+    try {
+      response = await api.uploadFile(auth, attachment.uri, fileName);
+    } catch (e) {
+      showToast(_('Uploading {fileName} failed.', { fileName }));
+      this.setMessageInputValue(
+        this.state.message.replace(
+          placeholder,
+          `[${_('Uploading {fileName} failed.', { fileName })}]()`,
+        ),
+      );
+      return;
+    }
+
+    const linkText = `[${fileName}](${response.uri})`;
+    if (this.state.message.indexOf(placeholder) !== -1) {
+      this.setMessageInputValue(this.state.message.replace(placeholder, linkText));
+    } else {
+      this.setMessageInputValue(`${this.state.message}\n${linkText}`);
+    }
+
+    this.setState(({ numUploading }) => ({
+      numUploading: numUploading - 1,
+    }));
```

Comment:
The new `return` in that `catch` block above causes me to notice a pitfall here: if anything in the function above this point exits the function early, this `numUploading` decrement won't run and we'll keep thinking there's an ongoing upload forever (until the user navigates out of the narrow so that this component gets discarded.)

In particular that will happen if the new `return` fires, but also in the previous revision it could happen if for some reason something inside that `catch` block (like the `this.setMessageInputValue` call, for example) were to throw an exception. (At the end of the branch, there's no longer the `return` but there's even more code in the function, so there are more opportunities for something to throw an exception.)

The way to fix this, in a way where it's easy to tell by looking at the code that it's definitely taken care of, is to put this decrement in a `finally` block. So the function might look like this:
```js
  insertAttachment = async (attachments: DocumentPickerResponse[]) => {
    this.setState(({ numUploading }) => ({
      numUploading: numUploading + 1,
    }));
    try {
      // …
    } finally {
      this.setState(({ numUploading }) => ({
        numUploading: numUploading - 1,
      }));
    }
  };
```

It's a bit annoying because it causes the whole rest of the function to be indented -- really the ideal pattern here would be [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization), like in C++ or Rust. But this is probably the best approximation to that that we have in JS.